### PR TITLE
Relative Auto Positioning

### PR DESCRIPTION
**Pull request by jshcrowthe**
### Description

_EDIT: The assumptions section has been updated from feedback_

This PR adds relative auto positioning to the dialog implementation. 

The positioning functionality assumes two things:

- The dialog being positioned will be the same size regardless of its location on the screen
- The dialog's parent containers **do not** have `overflow: hidden;`
- ~The dialog and it's "anchor" element (the element it is being positioned relative to), share the same `offsetParent` _(see https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetParent)_~

#### Potential Gotchas

In my implementation I forsee two potential gotchas:
- If the dialog isn't the same size regardless of screen location, the dialog may be positioned incorrectly.
- If the dialog has a parent with `overflow:hidden;` then, if positioned relative to an element, there is a possibility the dialog will not display properly.
- ~If the dialog and its anchor element don't share the same `offsetParent` then the position functionality will noop.~

#### Arrow Direction
The position functionality will change based on the existence of the `arrow-direction` attribute. 

**`arrow-direction` Exists**
If the attribute exists, the dialog will be centered such that the arrow points at the center of the anchor element. 

**`arrow-direction` Does NOT Exist**
If the attribute **does not** exist, the dialog will fly out to the side of the browser window with the most available space.

### CODE REVIEW

File: dialog-el.html [hunk 1]
```diff
@@ -296,7 +296,92 @@
...
+      this.style.top = 'initial';
+      this.style.right = 'initial';
+      this.style.bottom = 'initial';
+      this.style.left = 'initial';
+    }
+
+    proto.positionDialog = function(anchor) {
+      // If `anchor` is falsy OR the anchor and dialog don't share an
+      // offsetParent, early return
+      if (!anchor || anchor.offsetParent !== this.offsetParent) return;
```
Comment (Norman-Amy) [inline dialog-el.html]: Requiring the same offsetParent worries me a bit. How would that affect a global dialog?

Comment (liechtym) [reply dialog-el.html]: I have the same concern. I think it is not a safe assumption that the anchor element and the dialog will have the same offsetParent. It wouldn't work if I wanted to use one instance of dialog-el for all dialogs on the page.

File: index.html [hunk 2]
```diff
@@ -70,6 +70,41 @@
     <li><button class='arrowDirectionDialog'>Arrow Direction Dialog</button></li>
   </ul>
 
+  <h3>Auto Positioning Demo</h3>
+  <p>
+    <strong>NOTE:</strong> the auto positioning will only work <storng>reliably</storng>
```
Comment (liechtym) [inline index.html]: Typo: storng

File: index.html [hunk 3]
```diff
@@ -70,6 +70,41 @@
     <li><button class='arrowDirectionDialog'>Arrow Direction Dialog</button></li>
```
Comment (liechtym) [inline index.html:70]: This example isn't working well, I assume because the anchor and dialog it opens aren't offset parents.

**Review by liechtym — COMMENTED** [review]
This is a really clean implementation of auto positioning. The biggest complaint I have is that it is based on offset values, which require the same offsetParent. This makes it impossible to use this element as a global dialog.

**Comment by liechtym** [comment]
Here's further reasoning why using offset values for positioning isn't ideal. It sometimes requires that the dialog element is inserted far into the element tree. E.g., html > body > div > div > nav > ul > li > dialog-el, because you must put it near the triggering element so it has the same offset parent. If any of the parent tree elements has the CSS property `overflow: hidden`, the dialog will get cut off. It is more ideal to put the dialog element in the body tag, so as to avoid problems like this.

**Comment by liechtym** [comment]
I made a PR to this branch (https://github.com/jshcrowthe/dialog-el/pull/25) with an idea for positioning the dialog relative to the document. LMK what you think.

File: index.html (hunk 3 above)
Comment (jshcrowthe) [reply index.html:70]: That example isn't using the auto-position feature.

File: index.html (hunk 2 above)
Comment (jshcrowthe) [reply index.html]: Fixed in 56f6868

File: dialog-el.html (hunk 1 above)
Comment (jshcrowthe) [reply dialog-el.html]: Alright, I have pulled in @liechtym's PR (#25) and have made some modifications to adjust for variable element `offsetParent`. This should remove the need for this assumption.

However this introduces a different assumption: that the dialog has the potential to flow **outside** of its container due to auto-positioning. I am going to look into how I could create a separate stacking context for dialogs (CSS fixed hack anyone?) that could address this.

_Update: I don't think there is anything that will allow us to get around the `overflow:hidden;` issue. We'll have to just be explicit about this in the docs_

**Comment by liechtym** [comment]
I pulled down the latest of the auto-position branch, and it is looking awesome! I'm very happy with how it's coming along! Look forward to seeing this merged!

**Comment by liechtym** [comment]
I see the next step of this feature as including the ability for the component to adjust from it's non-default relative position if positioning it this way would make it appear partially or fully offscreen. I'll start thinking of possible simple implementations of this functionality.

**Comment by jshcrowthe** [comment]
Thanks @liechtym and @Norman-Amy for your feedback.

I will work on getting the tests fixed before I merge this. Unfortunately WCT is having some issues...

@liechtym to your comment on "next-steps for this feature," I would like to stay out of the business of what I'm calling "assuming user intent." What I mean by that is: I don't want to assume I know what the user means beyond what they have explicitly said. 

An example of this would be something like if I tried to render a modal instead of a dialog because of a screen size constraint. If the user says do something, I want to avoid side effects.

So in your prototyping try and take that into account. That said, I look forward to seeing what you come up with! Let's pull that discussion out into a separate issue so we can keep this PR thread clean.

**Comment by Norman-Amy** [comment]
We've been having trouble with WCT in our other components as well, would rolling back to an earlier version help?

From what I've seen while testing this, I'm not entirely satisfied. Resizing the window doesn't appear to change where the auto-positioned dialog goes, but moving the button that opens it does. Is that how it's supposed to work? Also, does the auto-positioning only work for non-arrow-direction dialogs? I know you don't want to interpret user intent by changing the default arrow direction so it will better fit on the screen, but I have to agree with @liechtym that it's a needed feature. In my mind, not doing that would defeat the whole purpose of auto-positioning in the first place. The user should be expecting it if they are aware the auto-positioning exists.

**Comment by jshcrowthe** [comment]
@Norman-Amy the issue I'm seeing with WCT isn't a rollback issue it's a screensize issue, it stems from an issue in `document.elementFromPoint()` (See: https://developer.mozilla.org/en-US/docs/Web/API/Document/elementFromPoint on elements not in visible viewport). The failing build is here: https://travis-ci.org/jshcrowthe/dialog-el/builds/202165112

Window resize is something that I had not taken into account! If a non-arrow-dialog is anchored to a button and a resize changes the quadrant the dialog should be rendered in, I could see a reposition being valuable. In that light, @liechtym's suggestion that we "smart-position" would make sense as we can optimize based on available viewport.

Auto Position supports arrow and non-arrow dialogs, from the above PR decription

> ### Arrow Direction
>
> The position functionality will change based on the existence of the arrow-direction attribute.
>
> **arrow-direction Exists**
> If the attribute exists, the dialog will be centered such that the arrow points at the center of the anchor element.
>
> **arrow-direction Does NOT Exist**
> If the attribute does not exist, the dialog will fly out to the side of the browser window with the most available space.

IMO arrow dialog positioning is less "smart." I don't think changing the `arrowDirection` property as part of the positioning step is a use case we want to get in to. I would envision smart positioning for an arrow-dialog being to optimize the arrow position. However I don't imagine us doing much else.

Thoughts?

**Comment by Norman-Amy** [comment]
I don't see why it can't be as smart. I mean as you said, when the window size changes, it needs to re-position and that would require moving the arrow along with it. In that case the one specified by the user becomes the default. Maybe I'm just not seeing the positioning work for the arrow dialog if it's already being applied. How do you test that?

**Comment by jshcrowthe** [comment]
My biggest sticking point on this is the following:

As a consumer, if I set the `arrowDirection` value, unless I specify I am flexible about arrow direction, it should be pointing the direction I say.

So, with that in mind, if we added a property that allowed a user to indicate that the `arrowDirection` prop is a "recommendation" rather than a "requirement," I would agree with you. That would be a separate feature, whereas the window resize and smart positioning for non-arrow-dialogs would be enhancements of the current feature.

Send me a JSBin with your use case. I can't tell what your problem is from your description. 

_(here is a boilerplate [JSBin](http://jsbin.com/xahaqu/4/edit?html,output) to get you started)_

**Comment by jshcrowthe** [comment]
@liechtym - I have finally got the PR tests green so we should be good to go. I have run all the tests locally and they pass consistently across all my browsers.

@Norman-Amy - Is there anything else from the comparison of the two implementations? I want to make sure we aren't forgetting anything.

**Comment by Norman-Amy** [comment]
I would be fine with adding a property like `allow-auto-position` that would specify if the user wants us to be able to move the arrow. Without that smart positioning, we won't be able to consume it. I can try to find some time to make a PR with that functionality as long as I know you're ok with it before-hand.

**Comment by liechtym** [comment]
Good to hear Josh! I'm glad all the tests are passing. Personally, I think this PR is good as is, and this "allowing auto positioning" idea would be best served as a separate PR.

I like the reasoning behind the idea that, if a developer says an arrow points in one direction, it stays. However, like Amy, I think this should be negotiable with an extra property, such as the suggested `allow-auto-position`. For my use cases, I would like to designate a certain direction I'd like the arrow to go, but if there is a screen size or orientation where my dialog has no room to display, I would prefer the dialog to auto adjust to another orientation. However, I can see other use cases where I would be annoyed if it auto adjusted when I didn't want it to. Giving me a property so I can choose its behavior would give me complete control over the dialog. What I'm saying is, I really like Amy's idea. But again, I feel that this is a separate feature that we should discuss in another issue, and merge the current one as is. This one is a good stepping stone in the right direction for what I'd be looking for in using this dialog.

**Comment by liechtym** [comment]
@jshcrowthe When will this be merged?

**Pull request merged**